src/shared/tests.c

```diff
@@ -301,3 +301,26 @@ int enter_cgroup_subroot(char **ret_cgroup) {
 int enter_cgroup_root(char **ret_cgroup) {
         return enter_cgroup(ret_cgroup, false);
 }
+
+const char *ci_environment(void) {
+        /* We return a string because we might want to provide multiple bits of information later on: not
+         * just the general CI environment type, but also whether we're sanitizing or not, etc. The caller is
+         * expected to use strstr on the returned value. */
+        static const char *ans = NULL;
+
+        if (ans)
+                return ans;
+
+        if (streq_ptr(getenv("TRAVIS"), "true"))
+                return (ans = "travis");
+        if (streq_ptr(getenv("SEMAPHORE"), "true"))
+                return (ans = "semaphore");
+        if (streq_ptr(getenv("GITHUB_ACTIONS"), "true"))
+                return (ans = "github-actions");
+        if (getenv("AUTOPKGTEST_ARTIFACTS") || getenv("AUTOPKGTEST_TMP"))
+                return (ans = "autopkgtest");
+        if (getenv("CI") || getenv("CONTINOUS_INTEGRATION"))
+                return (ans = "unknown"); /* Some other unknown thing */
+
+        return (ans = "");
```

Comment:
this is weird, no? shouldn't this return NULL?

Reply:
and then use ((char*) -1) as cached value that means "untested"?

Reply:
I think this gives nice semantics on the test side. If -1 was returned this would require a more complicated check, since -1 cannot be dereferenced successfully. Again, this is not supposed to be a general-purpose API, but just something that we can use internally. If it turns out we need something more complicated, we can always replace this in the future.

Reply:
i didn't mean return -1, but return NULL. (but internally in the cache var use -1 to distinguish "cache not set yet" from "no CI env")